### ISSUE TITLE
SPI intermittent output failure

### ISSUE THREAD

**Issue by dleibrandt** [m-labs/artiq#672]
I'm running artiq 2.2 on the kc705 nist clock hardware.

I'm testing SPI communication with a slave device by writing random values to configuration registers and reading them back, and I see intermittent failures of the SPI RTIO outputs for SPI clock frequencies faster than about 2 MHz.  The rate of these failures gets higher as the SPI clock frequency is increased.  Below 2 MHz, I never see failures (in > 1e5 trials).  Below are a few example scope traces.  Yellow is data in, blue is data out, purple is chip select out, and green is clock out.

Successful trial:
![img_0660](https://cloud.githubusercontent.com/assets/13787968/23047430/f291b194-f46c-11e6-968d-c14a72407ffe.JPG)

Failed trial 1 (the final few clock cycles were not output):
![img_0664](https://cloud.githubusercontent.com/assets/13787968/23047436/f9ab9eea-f46c-11e6-83f1-484dd2eb2ce1.JPG)

Failed trial 2 (the final few clock cycles were not output and chip select was late to be pulled high):
![img_0662](https://cloud.githubusercontent.com/assets/13787968/23047438/febc1694-f46c-11e6-9ded-c97565d458e2.JPG)

Failed trial 3 (similar mode):
![img_0663](https://cloud.githubusercontent.com/assets/13787968/23047441/032d37f8-f46d-11e6-9110-c6261db565bf.JPG)

**Comment by sbourdeauducq** [comment]
Can you post kernel code that produces this issue?

**Comment by dleibrandt** [comment]
Something like this:

```
    @kernel
    def run(self):
        self.core.reset()
        delay(10*ms)

        self.setup_bus()
        delay(10*ms)

        self.servo._get_config_uint(addr)

class DigitalServo:
    """
    SPI communication with NIST digital servos.

    :param spi_device: Name of the SPI bus this device is on.
    :param sample_device: Name of the TTL device that the is connected to
      the sample input (DIN2) of the digital servo (optional).
    """

    def __init__(self, dmgr, spi_device, sample_device=None):
        self.core = dmgr.get("core")
        self.bus = dmgr.get(spi_device)
        if sample_device is not None:
            self.sample = dmgr.get(sample_device)
        
        self.address = 0

    @kernel
    def setup_bus(self, write_div=50, read_div=50):
        self.bus.set_config_mu(_DIGITAL_SERVO_SPI_CONFIG, write_div, read_div)
        self.bus.set_xfer(1, 16, 16)
    
    @kernel
    def _get_config_uint(self, address):
        self.bus.write(((self.address & 0x3) << 30) | (0x1 << 28) | ((address & 0xFFF) << 16))
        self.bus.read_async()
        
        return int(self.bus.input_async() & 0xFFFF)

```

**Comment by jordens** [comment]
I am worried about that `core.reset()`.
Could you provide an analyzer dump? I think the SPI channels should be covered by coreanalyzer but haven't checked lately.

**Comment by sbourdeauducq** [comment]
There will be in the raw dump but decoding into VCD is not supported. @dleibrandt please provide the raw dump.

**Comment by dleibrandt** [comment]
I've attached both below.

[rtio.zip](https://github.com/m-labs/artiq/files/783927/rtio.zip)

**Comment by jordens** [comment]
@dleibrandt thanks.
* Did that dump include such an intermittent failure?
* What do you need the `core.reset()` for? Could you try without it (e.g. `break_realtime()` where needed instead)?
* The "intermittent" baffles me. I can't come up with a reason based on your description and your code that could make something like this intermittent.
* Your dump shows that there are `read_async()` immediately followed by `set_config_mu()`/`set_xfer()`. That's not how you do it in your example code. If you reset/change `config` or `xfer` or start a new `write` before the `read` has finished, or if you `core.reset()` the entire thing, that would obviously cause problems. Please provide the code that shows the problem and the corresponding dump.

I have added SPI support to coreanalyzer (in master).

**Comment by dleibrandt** [comment]
> Did that dump include such an intermittent failure?

Yes.

> What do you need the core.reset() for? Could you try without it (e.g. break_realtime() where needed instead)?

You're right.  I don't actually need the `core.reset()`.  However, I still see the problem when I replace it with `core.break_realtime()`.

> The "intermittent" baffles me. I can't come up with a reason based on your description and your code that could make something like this intermittent.

Agreed.  I also don't have an explanation.

> Your dump shows that there are read_async() immediately followed by set_config_mu()/set_xfer(). That's not how you do it in your example code. If you reset/change config or xfer or start a new write before the read has finished, or if you core.reset() the entire thing, that would obviously cause problems. Please provide the code that shows the problem and the corresponding dump.

Ah, yes.  Sorry about that.  I was trying to simplify things a bit, and I've been changing the code a lot lately.  This code:
```
from artiq.experiment import *
import time

class TestDigitalServo(EnvExperiment):
    def build(self):
        self.setattr_device("core")
        self.setattr_device("scheduler")
        
        self.setattr_device("servo0")
        self.setattr_device("servo1")
        
        self.setattr_device("ttl12")
    
    def run(self):
        self.servo0.test_spi_com()



from artiq.language.core import (kernel, portable, delay_mu, delay)
from artiq.language.units import ns, us, ms
from artiq.language.types import TInt64, TInt32, TNone, TList
from artiq.coredevice import spi

_DIGITAL_SERVO_SPI_CONFIG = (0*spi.SPI_OFFLINE | 0*spi.SPI_CS_POLARITY |
                             0*spi.SPI_CLK_POLARITY | 0*spi.SPI_CLK_PHASE |
                             0*spi.SPI_LSB_FIRST | 0*spi.SPI_HALF_DUPLEX)

class DigitalServo:
    """
    SPI communication with NIST digital servos.

    :param spi_device: Name of the SPI bus this device is on.
    :param address: SPI address (used for daisy chaining multiple boxes).
    :param sample_device: Name of the TTL device that the is connected to
      the sample input (DIN2) of the digital servo (optional).
    """

    def __init__(self, dmgr, spi_device, address, sample_device=None):
        self.core = dmgr.get("core")
        self.bus = dmgr.get(spi_device)
        self.address = address
        if sample_device is not None:
            self.sample = dmgr.get(sample_device)

    @kernel
    def setup_bus(self, write_div=25, read_div=50): # write_div=25, read_div=125):
        """Configure the SPI bus and the SPI transaction parameters
        for this device. This method has to be called before any other method
        if the bus has been used to access a different device in the meantime.

        This method advances the timeline by the duration of two
        RTIO-to-Wishbone bus transactions.

        :param write_div: Write clock divider.
        :param read_div: Read clock divider.
        """
        
        self.bus.set_config_mu(_DIGITAL_SERVO_SPI_CONFIG, write_div, read_div)
        self.bus.set_xfer(1, 32, 0)

    @kernel
    def _set_config(self, address, value):
        """Write a 16 bit configuration register.

        This method advances the timeline by the duration of the SPI transfer
        and the required CS high time.
        
        :param address: Address.
        :param value: Value.
        """
        
        self.bus.write(((self.address & 0x3) << 30) | (0x0 << 28) | ((address & 0xFFF) << 16) | (value & 0xFFFF))

    @kernel
    def _get_config_uint(self, address):
        """Read a 16 bit signal register.

        This method advances the timeline by the duration of :meth:`write`.

        :param address: Address.
        """
        
        self.bus.set_xfer(1, 16, 16)
        
        self.bus.write(((self.address & 0x3) << 30) | (0x1 << 28) | ((address & 0xFFF) << 16))
        self.bus.read_async()
        
        self.bus.set_xfer(1, 32, 0)
        
        return int(self.bus.input_async() & 0xFFFF)

    @kernel
    def _get_signal(self, address, n_bytes = 4):
        """Read a 32 bit signal register.

        This method advances the timeline by the duration of :meth:`write`.

        :param address: Address.
        """
        
        self.bus.set_xfer(1, 16, 8*n_bytes)
        
        self.bus.write(((self.address & 0x3) << 30) | (0x2 << 28) | ((address & 0x7F) << 16))
        self.bus.read_async()
        
        self.bus.set_xfer(1, 32, 0)
        
        return self.bus.input_async()

    def test_spi_com(self):
        self.n_trials = 1000
        
        # generate a list of random uint16's
        self.x = [np.random.randint(0, 2**16) for _ in range(self.n_trials)]
        
        self._test_spi_com()
        
    @kernel
    def _test_spi_com(self):
        """Test if spi communication is working.
        
        Returns true if all tests pass; false otherwise.
        """
        
        pass_test = True
        
        # reset and setup
        #self.core.reset()
        self.core.break_realtime()
        delay(10*ms)

        self.setup_bus()
        delay(10*ms)
        
        # test _get_signal()
        print("Testing _get_signal() ...")
        n_failed = 0
        for i in range(self.n_trials):
            value = self._get_signal(35)
            
            if ((value & 0xFFFFFFFF) != 0xE80F00FF):
                n_failed += 1
                pass_test = False
            
            delay(10*us)
        
        if (n_failed > 0):
            print("    failed", n_failed, "out of", self.n_trials, "trials")
        else:
            print("    passed (", self.n_trials, " trials)")
        delay(10*ms)
        
        # test _set_config() and _get_config_uint()
        print("Testing _set_config() and _get_config_uint() ...")
        n_failed = 0
        for i in range(self.n_trials):
            n_config_words = 768
            addr = (i % (n_config_words - 1)) + 1 # don't write to word 0 (spi address)
            
            v1 = self.x[i]
            
            self._set_config(addr, v1)
            delay(100*us)
            v2 = self._get_config_uint(addr)
            
            delay(5*us)
            
            if (v1 != v2):
                self.sample.pulse(10*us)
                
                n_failed += 1
                pass_test = False
                
                print("    failed: trial =", i, ", set =", v1, ", get =", v2)
                delay(10*ms)
        
        if (n_failed > 0):
            print("    failed", n_failed, "out of", self.n_trials, "trials")
        else:
            print("    passed (", self.n_trials, " trials)")
        delay(10*ms)
        
        return pass_test
```

produces the problem on 14% of attempts.  Adding 1 ms delays on both sides of the `set_xfer` calls in `_get_config_uint` reduces that to 8%.  Increasing `read_div` to 125 eliminates the problem altogether.  Does `set_xfer` not respect the timeline?

In any case, the exact code above produced the RTIO dump below, which failed on the last attempt.

[rtio.zip](https://github.com/m-labs/artiq/files/794779/rtio.zip)

Here is the corresponding log output:
```
INFO:worker(6108,TestDigitalServo.py):print:Testing _get_signal() ...
INFO:worker(6108,TestDigitalServo.py):print:    passed ( 1000  trials)
INFO:worker(6108,TestDigitalServo.py):print:Testing _set_config() and _get_config_uint() ...
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 15 , set = 47757 , get = 47744
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 20 , set = 19112 , get = 19132
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 46 , set = 36518 , get = 36519
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 52 , set = 27294 , get = 27264
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 55 , set = 16324 , get = 22552
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 60 , set = 5451 , get = 5440
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 69 , set = 31154 , get = 50095
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 70 , set = 43537 , get = 43520
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 89 , set = 63709 , get = 36724
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 91 , set = 54534 , get = 54528
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 103 , set = 56309 , get = 7262
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 135 , set = 58084 , get = 40606
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 147 , set = 1020 , get = 20519
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 150 , set = 51858 , get = 51868
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 153 , set = 53898 , get = 53903
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 164 , set = 21164 , get = 21152
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 183 , set = 38228 , get = 38208
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 187 , set = 1364 , get = 1367
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 197 , set = 11947 , get = 11944
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 217 , set = 13976 , get = 28132
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 221 , set = 31644 , get = 26262
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 240 , set = 43668 , get = 43775
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 254 , set = 7753 , get = 33028
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 268 , set = 40361 , get = 40360
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 280 , set = 34902 , get = 935
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 282 , set = 10909 , get = 10880
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 302 , set = 38308 , get = 38311
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 304 , set = 54339 , get = 16275
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 308 , set = 21721 , get = 12736
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 327 , set = 54771 , get = 54783
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 329 , set = 8875 , get = 57189
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 336 , set = 15598 , get = 22406
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 344 , set = 4196 , get = 16803
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 352 , set = 28226 , get = 24113
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 366 , set = 12980 , get = 60145
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 373 , set = 4771 , get = 4768
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 383 , set = 24304 , get = 50307
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 387 , set = 45639 , get = 56364
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 389 , set = 56148 , get = 56151
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 398 , set = 24694 , get = 55075
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 400 , set = 19644 , get = 7455
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 402 , set = 18839 , get = 20407
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 419 , set = 36811 , get = 63888
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 440 , set = 4439 , get = 18181
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 444 , set = 43631 , get = 43520
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 452 , set = 41352 , get = 16714
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 468 , set = 17758 , get = 17759
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 480 , set = 22168 , get = 22175
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 502 , set = 8534 , get = 8535
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 507 , set = 59099 , get = 20069
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 513 , set = 51868 , get = 51871
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 518 , set = 44717 , get = 44719
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 546 , set = 55978 , get = 55999
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 553 , set = 56659 , get = 56656
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 561 , set = 37600 , get = 16383
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 568 , set = 58697 , get = 58689
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 571 , set = 6829 , get = 6831
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 575 , set = 45153 , get = 25462
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 578 , set = 16307 , get = 62552
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 580 , set = 22226 , get = 22227
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 584 , set = 19934 , get = 10408
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 593 , set = 60036 , get = 60032
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 609 , set = 64081 , get = 64080
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 616 , set = 19167 , get = 6345
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 623 , set = 9183 , get = 15736
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 628 , set = 38768 , get = 2413
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 631 , set = 38307 , get = 38304
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 633 , set = 36244 , get = 10234
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 634 , set = 59722 , get = 59720
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 661 , set = 50340 , get = 25142
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 668 , set = 27996 , get = 27999
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 677 , set = 53905 , get = 53904
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 699 , set = 7291 , get = 18246
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 707 , set = 3468 , get = 28855
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 713 , set = 51203 , get = 14623
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 727 , set = 35111 , get = 11493
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 736 , set = 21784 , get = 21760
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 760 , set = 35483 , get = 35487
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 766 , set = 43563 , get = 43520
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 769 , set = 43787 , get = 7318
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 772 , set = 53936 , get = 53948
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 797 , set = 34619 , get = 61470
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 800 , set = 40618 , get = 40616
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 801 , set = 55937 , get = 55936
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 802 , set = 30373 , get = 30375
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 804 , set = 9550 , get = 9551
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 815 , set = 8873 , get = 8872
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 819 , set = 43650 , get = 27264
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 841 , set = 5546 , get = 5544
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 858 , set = 52907 , get = 52904
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 863 , set = 41813 , get = 41812
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 864 , set = 36166 , get = 36161
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 881 , set = 33343 , get = 22941
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 882 , set = 42402 , get = 42400
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 895 , set = 60742 , get = 60736
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 900 , set = 46531 , get = 27643
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 924 , set = 38238 , get = 38271
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 933 , set = 63636 , get = 51547
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 961 , set = 31427 , get = 18251
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 976 , set = 21929 , get = 21928
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 981 , set = 41081 , get = 42131
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 989 , set = 58793 , get = 58792
INFO:worker(6108,TestDigitalServo.py):print:    failed: trial = 999 , set = 9045 , get = 9044
INFO:worker(6108,TestDigitalServo.py):print:    failed 103 out of 1000 trials
```

And here is the physical scope output:

![img_0665](https://cloud.githubusercontent.com/assets/13787968/23232372/3968e08e-f908-11e6-8346-99111e717cf2.JPG)

**Comment by jordens** [comment]
@dleibrandt just a quick note to confirm that I am still looking at this. Right now, the only things I can come up with is some weirdness or deep bug in the gateware or in Vivado that we haven't seen before or a spurious RTIO reset interfering.

**Comment by sbourdeauducq** [comment]
Did anyone make any progress?

**Comment by jordens** [comment]
I ran SPI reads over night on PDQ and couldn't reproduce this.

**Comment by dleibrandt** [comment]
Using the 2.2 nist_clock gateware, or a different gateware?

**Comment by sbourdeauducq** [comment]
I suppose 3.0/master, but the SPI code should be the same.

**Comment by jordens** [comment]
Yes.

**Comment by sbourdeauducq** [comment]
@dleibrandt Do you still have the problem with 2.3 (instead of 2.2) and that commit Robert mentions?

**Comment by sbourdeauducq** [comment]
@dleibrandt Also it would be helpful if you found the simplest possible experiment code that produces the problem.

**Comment by jordens** [comment]
With current master, I ran a couple million of these 16 MHz reads and they all passed fine.

```python
from artiq.experiment import *
from artiq.coredevice import spi

_DIGITAL_SERVO_SPI_CONFIG = (0*spi.SPI_OFFLINE | 0*spi.SPI_CS_POLARITY |
                             0*spi.SPI_CLK_POLARITY | 0*spi.SPI_CLK_PHASE |
                             0*spi.SPI_LSB_FIRST | 0*spi.SPI_HALF_DUPLEX)

class Test(EnvExperiment):
    def build(self):
        self.setattr_device("core")
        self.setattr_device("spi_sma")
        self.setattr_device("led")

    def run(self):
        self.q(10000000)

    @kernel
    def q(self, n):
        self.core.reset()
        self.core.break_realtime()

        self.spi_sma.set_config_mu(flags=_DIGITAL_SERVO_SPI_CONFIG,
                write_div=8, read_div=8)
        self.spi_sma.set_xfer(chip_select=0b1, write_length=16, read_length=16)

        for i in range(n):
            self.led.on()
            self.spi_sma.write(0x8001)
            delay(20*ns)
            self.led.off()
            self.spi_sma.read_sync()
            delay(10*us)
```

**Comment by dleibrandt** [comment]
Just checked 2.3, and I still see the problem.

**Comment by sbourdeauducq** [comment]
Can you test with master, and post a minimal way to reproduce the problem?

**Comment by dleibrandt** [comment]
Spent some time trying to come up with a simple way to reproduce the problem, and learned that it is only present when the SPI lines are connected to my slave device.  Further testing revealed that the SPI lines coming out of the kc705 were actually OK, but previously I had only measured the SPI lines after an isolator chip on the TTL breakout board.  The isolator chip outputs were dropping out because the MISO line was ringing (overshooting), causing the voltage to be outside the specs of the chip.  I fixed this by reducing the MISO drive current compliance on the slave device.

Anyway, problem solved.  Sorry about the wild goose chase.
